src/QsCompiler/CompilationManager/CompilationUnitManager.cs

```diff
@@ -68,12 +82,40 @@ public CompilationUnitManager(Action<Exception> exceptionLogger = null, Action<P
             this.CompilationUnit = new CompilationUnit();
             this.FileContentManagers = new ConcurrentDictionary<NonNullable<string>, FileContentManager>();
             this.ChangedFiles = new ManagedHashSet<NonNullable<string>>(new ReaderWriterLockSlim(LockRecursionPolicy.SupportsRecursion));
-            this.PublishDiagnostics = publishDiagnostics ?? (_ => { });
+            this.PublishDiagnostics = diagnostic =>
+            {
+                lock (this.PendingDiagnostics)
```

Comment:
Hmm, semi I think. I am not sure this is a good option, since it then may pick up partially computed diagnostics. At the end the diagnostics would be correct, but I am not sure if the changes while typing might make for a bad experience.